### PR TITLE
Separated GitCheckout into Github and Stash Downloaders, and added stash DL support

### PR DESCRIPTION
**Pull request by zanchin**

### CODE REVIEW

File: defaults.yaml [hunk 1]
```diff
@@ -68,3 +68,7 @@ loom:
 # asset server (must NOT be localhost because it's called from within container)
 assets:
   url: "http://localhost:3070"
+
+# auth credentials service (i.e., for stash)
+auth:
```
Comment (tizzo) [inline defaults.yaml]: Can this work with anything other than stash? If not shouldn't this be called stash config? Maybe grouped under the configuration for the stash plugin specifically?

File: lib/Container.js [hunk 2]
```diff
@@ -63,12 +58,38 @@ Container.prototype.buildSetupTasks = Promise.promisify(function(done) {
...
 
+Container.prototype._createDownloader = function(build, project){
+  // map provider type to downloader class name
+  var downloaders = {
+    'github': 'GithubDownloader',
+    'stash': 'StashDownloader',
+  }
+
+  var provider_type = project.provider.type
+  var clazz = downloaders[provider_type]
```
Comment (tizzo) [inline lib/Container.js]: Why not use a real word here like `plugin` rather than a slight WTF like `clazz`?

File: lib/Container.js [hunk 3]
```diff
@@ -63,12 +58,38 @@ Container.prototype.buildSetupTasks = Promise.promisify(function(done) {
...
+    throw new Error("Download: unsupported provider type: " + provider_type)
+  }
+
+  // dynamically require and instantiate a downloder
+  var Downloader
+  try {
+    Downloader = require(`./plugins/TaskRunner/${clazz}`)
+  } catch (e) {
+    var msg = `Download: failed to load supported downloader for provider ${provider_type}: ${e.message}`
+    this.log.error({err: e, clazz}, msg)
```
Comment (tizzo) [inline lib/Container.js]: We shouldn't lean implementation details like the `clazz` variable name into the error log.

File: lib/Container.js [hunk 4]
```diff
@@ -138,9 +159,6 @@ Container.prototype.create = Promise.promisify(function(done) {
       `PWD=${SRC_DIR}`,
     ],
   }
-  if (this.build.ref) {
-    createOptions.Env.push(`COMMIT_REF=${this.build.ref}`);
```
Comment (tizzo) [inline lib/Container.js]: Is this also defined elsewhere or are we dropping a variable?

File: defaults.yaml (hunk 1 above)
Comment (zanchin) [inline defaults.yaml]: Any OAuth1 service could need that, so keeping the interface generic. That's why this endpoint returns the full value of the `Authorization` header to keep implementation details abstracted.

File: lib/Container.js (hunk 2 above)
Comment (zanchin) [inline lib/Container.js]: Ok, will do.

File: lib/Container.js (hunk 3 above)
Comment (zanchin) [inline lib/Container.js]: Solved by renaming `clazz`->`plugin`

File: lib/Container.js (hunk 4 above)
Comment (zanchin) [inline lib/Container.js]: Yeah, it's already defined on line [#134](https://github.com/ProboCI/probo/pull/8/files#diff-599aa5690d4bc1ef4f297bfe52690761L134).

Comment (tizzo) [inline lib/Container.js]: Durp, missed that. Thanks!

**Pull request merged**